# Introduction to Linux - March 2023

The purpose of this short course is to teach researchers how to use the Linux operating system, thereby enhancing their ability to utilize various types of bioinformatics tools and support their research efforts. 

# Schedule 

Four workshops (total 10 hours)

Slatyer Room (Rm N2011) - Robertson Building #46 

* Tuesday, 28 Feb 1.30-4.00pm
* Tuesday, 7 Mar 1.30-4.00pm
* Tuesday, 14 Mar 1.30-4.00pm
* Tuesday, 21 Mar 1.30-4.00pm 

For each workshop, the class schedule is:

* 1.30-2.20pm (50min) Teaching 
* 2.20-2.40pm (20min) Break and discussion 
* 2.40-3.30pm (50min) Teaching 
* 3.30-3.50pm (20min) Break and discussion 
* 3.50-4.00pm (10min) Wrap up

This course is also available online via [Zoom](https://anu.zoom.us/j/83492102360?pwd=akdBdXI2SCtkdGg1S09ZY0NYTTQxQT09). 

# Course Contents 

## Week 1 - Basic Linux commands 

* Connecting to a remote server
* Navigating through directories 
* Creating and editing files 
* Viewing file contents 
* Moving, copying, deleting, and renaming files and directories
* Getting help with a command 
* Linux file paths and structures 
* Using shortcuts in Linux 
* Using tmux to keep running your sessions in the background 

## Week 2 - Basic Linux commands 

* Piping and redirecting standard input and output 
* Counting, sorting, and shuffling texts and file contents 
* Finding and replacing texts 
* Downloading and transfering data 
* Using loops and statements

## Week 3 - Variant calling workflow 

* Set up conda environment
* Assessing read quality using FastQC 
* Quality control 
* Trimming and filtering using trimmomatic 
* Aligning reads to reference genome using BWA 
* Variant calling using bcftools

## Week 4 - Automating a variant calling workflow 

* Writing a bash script about the workflow 
* Using slurm to run jobs on the server 
* Writing a sbatch file 
* Parallel processing 